### PR TITLE
[ Conv2D ] add cblas_sgemm to calculate conv2d

### PR DESCRIPTION
**Pull request by jijoongmoon**
This PR includes conv2d layer gradient calculation using cblas_gemm.
In order to do this, new functions are introduced.
. conv2d_gemm
. im2col

related issue : #477

Signed-off-by: jijoongmoon <jijoong.moon@samsung.com>

### CODE REVIEW

File: test/unittest/unittest_nntrainer_layers.cpp
Comment (kparichay) [inline test/unittest/unittest_nntrainer_layers.cpp:942]: How about add this unittest in a different PR.

File: test/unittest/unittest_nntrainer_layers.cpp
Comment (kparichay) [inline test/unittest/unittest_nntrainer_layers.cpp:1011]: How about add this unittest in a different PR.

File: test/unittest/unittest_nntrainer_layers.cpp
Comment (kparichay) [inline test/unittest/unittest_nntrainer_layers.cpp:1147]: ditto.

File: test/unittest/unittest_nntrainer_layers.cpp
Comment (kparichay) [reply test/unittest/unittest_nntrainer_layers.cpp:1011]: also, updating the unittest_layers.tar.gz with integers still needs more testing with loss. Lets do that at once in a separate PR.

File: nntrainer/src/conv2d_layer.cpp
Comment (kparichay) [inline nntrainer/src/conv2d_layer.cpp:529]: it seems like conv2d_gemm is only used in USE_BLAS mode. How about declaring and defining it under `#if defined(USE_BLAS) , #endlf`?
because this function isn't that useful without, neither its used without blas.

File: nntrainer/src/conv2d_layer.cpp
Comment (kparichay) [inline nntrainer/src/conv2d_layer.cpp:211]: I recommend you to tensor.sum_by_batch() for this. It uses blas internally if available and will be faster.

File: nntrainer/src/conv2d_layer.cpp
Comment (kparichay) [inline nntrainer/src/conv2d_layer.cpp:218]: ditto.

File: nntrainer/src/conv2d_layer.cpp
Comment (kparichay) [inline nntrainer/src/conv2d_layer.cpp:232]: can you put small comments. I am not sure section what this does.

File: nntrainer/src/conv2d_layer.cpp
Comment (kparichay) [inline nntrainer/src/conv2d_layer.cpp:256]: how about putting  `ret.getAddress(b * ret.getDim().getFeatureLen())` at line 252 directly in conv2d_gemm.
I see no reason to get the result in out and then make another copy.

File: test/unittest/unittest_nntrainer_layers.cpp
Comment (kparichay) [reply test/unittest/unittest_nntrainer_layers.cpp:1011]: Further corresponding gen_input.py also needs proper changes to be able to produce this unittest data.
If you want to add it yourself, please refer to https://github.com/kparichay/nntrainer/tree/conv_forwarding_mismatch
, or I will add this unittest after this PR is merged.

Comment (jijoongmoon) [reply test/unittest/unittest_nntrainer_layers.cpp:1011]: I will.

File: nntrainer/src/conv2d_layer.cpp
Comment (jijoongmoon) [reply nntrainer/src/conv2d_layer.cpp:529]: Well, there will be three way to calculate conv2d layer. one with sgemm ( blas & manual for matrix multiplication #491  ) and previous implementation without blas like. I think the previous version could be deleted I guess.

File: nntrainer/src/conv2d_layer.cpp
Comment (jijoongmoon) [reply nntrainer/src/conv2d_layer.cpp:211]: out is not tensor. It just is vector. There might be copy and I'm not sure it is faster.

File: nntrainer/src/conv2d_layer.cpp
Comment (jijoongmoon) [reply nntrainer/src/conv2d_layer.cpp:218]: I will fix later PR. This PR has too much already.

File: nntrainer/src/conv2d_layer.cpp
Comment (jijoongmoon) [reply nntrainer/src/conv2d_layer.cpp:232]: Could you give me more specific about section?

File: nntrainer/src/conv2d_layer.cpp
Comment (jijoongmoon) [reply nntrainer/src/conv2d_layer.cpp:256]: fixed. Thanks.

File: nntrainer/src/conv2d_layer.cpp
Comment (jijoongmoon) [reply nntrainer/src/conv2d_layer.cpp:218]: fixed. Thanks.

File: nntrainer/src/conv2d_layer.cpp
Comment (kparichay) [reply nntrainer/src/conv2d_layer.cpp:218]: you consider cblas_gemv here then (as this code is for blas only).

File: nntrainer/src/conv2d_layer.cpp
Comment (zhoonit) [inline nntrainer/src/conv2d_layer.cpp:254]: This constructor does memcpy. I think no need to memcpy here if conv2d_gemm takes argument of const float * in instead of const const Tensor & in

File: nntrainer/src/conv2d_layer.cpp
Comment (zhoonit) [inline nntrainer/src/conv2d_layer.cpp:499]: I think `in` does not need to be tensor rather float * as stated above

File: nntrainer/src/conv2d_layer.cpp
Comment (zhoonit) [inline nntrainer/src/conv2d_layer.cpp:505]: IMO `inCol` does not need to be vector and name should be `in_col`

File: nntrainer/src/conv2d_layer.cpp
Comment (zhoonit) [inline nntrainer/src/conv2d_layer.cpp:538]: rdata does nothing. I think we can just use `out` here.

File: nntrainer/src/conv2d_layer.cpp
Comment (zhoonit) [inline nntrainer/src/conv2d_layer.cpp:546]: Considering that this function is fairly complex, I think we should have a test for this function.

File: nntrainer/src/conv2d_layer.cpp
Comment (zhoonit) [inline nntrainer/src/conv2d_layer.cpp:515]: I guess im2col can throw instead

File: nntrainer/src/conv2d_layer.cpp
Comment (zhoonit) [inline nntrainer/src/conv2d_layer.cpp:264]: IMO, conv2d_gemm can throw with void signature and we do not need to rethrow again.

File: nntrainer/include/conv2d_layer.h
Comment (zhoonit) [inline nntrainer/include/conv2d_layer.h:152]:
```suggestion
   * @param[out] out output data
```

File: nntrainer/src/conv2d_layer.cpp
Comment (zhoonit) [inline nntrainer/src/conv2d_layer.cpp:109]: Do we need to copy data to `imKernel`?

File: nntrainer/src/conv2d_layer.cpp
Comment (kparichay) [reply nntrainer/src/conv2d_layer.cpp:536]: yes.

File: nntrainer/src/conv2d_layer.cpp
Comment (kparichay) [reply nntrainer/src/conv2d_layer.cpp:218]: You can directly use blas here as it is under USE_BLAS case. Then there will not be any copy.

File: nntrainer/src/conv2d_layer.cpp
Comment (kparichay) [reply nntrainer/src/conv2d_layer.cpp:225]: very strange. Ideally blas one should have been faster as it eliminates all the loops (even the one for batch)

File: nntrainer/src/conv2d_layer.cpp
Comment (kparichay) [inline nntrainer/src/conv2d_layer.cpp:264]: Backwardinf convolution failed??

File: nntrainer/include/conv2d_layer.h
Comment (jijoongmoon) [reply nntrainer/include/conv2d_layer.h:146]: m_kernel?? if you consider capital K, then I will change it mkernel then.

File: nntrainer/src/conv2d_layer.cpp
Comment (jijoongmoon) [reply nntrainer/src/conv2d_layer.cpp:109]: yes. we need.

File: nntrainer/src/conv2d_layer.cpp
Comment (jijoongmoon) [reply nntrainer/src/conv2d_layer.cpp:114]: yes. we need.

File: nntrainer/src/conv2d_layer.cpp
Comment (jijoongmoon) [reply nntrainer/src/conv2d_layer.cpp:218]: I will change with blas func. later.

File: nntrainer/src/conv2d_layer.cpp
Comment (jijoongmoon) [reply nntrainer/src/conv2d_layer.cpp:254]: the input of conv2d_gemm is tensor which is with dimension and it is used to calculate padding.

File: nntrainer/src/conv2d_layer.cpp
Comment (jijoongmoon) [reply nntrainer/src/conv2d_layer.cpp:264]: update later.

File: nntrainer/src/conv2d_layer.cpp
Comment (jijoongmoon) [reply nntrainer/src/conv2d_layer.cpp:499]: needed to calculate padding.

File: nntrainer/src/conv2d_layer.cpp
Comment (jijoongmoon) [reply nntrainer/src/conv2d_layer.cpp:505]: fixed. thanks.

File: nntrainer/src/conv2d_layer.cpp
Comment (jijoongmoon) [reply nntrainer/src/conv2d_layer.cpp:546]: another pr I guess. ( Besides, if this is wrong, all the test cases releated with conv2d are failed )

File: nntrainer/include/conv2d_layer.h
Comment (jijoongmoon) [reply nntrainer/include/conv2d_layer.h:189]: fixed. thanks.

File: nntrainer/include/conv2d_layer.h
Comment (jijoongmoon) [reply nntrainer/include/conv2d_layer.h:152]: updated. thanks.

File: nntrainer/src/conv2d_layer.cpp
Comment (zhoonit) [reply nntrainer/src/conv2d_layer.cpp]: If that's the case, I guess padding can be done outside of conv2d to avoid memcpy. I think it could be done later.

File: nntrainer/src/conv2d_layer.cpp
Comment (jijoongmoon) [reply nntrainer/src/conv2d_layer.cpp]: fixed. thanks.

File: nntrainer/src/conv2d_layer.cpp
Comment (jijoongmoon) [reply nntrainer/src/conv2d_layer.cpp:516]: Another PR I guess. Thanks.

**Review by zhoonit — APPROVED** [review]
LGTM after all comment has been resolved.

**Review by kparichay — APPROVED** [review]
LGTM
lets do the minor changes in the next PR

**Pull request merged**
